sycl/include/sycl/half_type.hpp

```diff
@@ -137,7 +137,8 @@ namespace host_half_impl {
 // The main host half class
 class __SYCL_EXPORT half {
 public:
-  half() = default;
+  constexpr half() = default;
```

Comment:
Note for reviewers: this is a non-trivial constructor! `std::is_trivially_default_constructible_v<sycl::half>` is `false`.

Reason for that is because trivially constructible classes can't have non-static members with default initializers. From [cppreference](https://en.cppreference.com/w/cpp/language/default_constructor):

> **Trivial default constructor**
> The default constructor for class `T` is trivial (i.e. performs no action) if all of the following is true:
> - The constructor is not user-provided (i.e., is implicitly-defined or defaulted on its first declaration)
> - `T` has no non-static members with default initializers. (_since C++11_)

We have non-static member `Buf` with default initializer below and we can't remove it or otherwise we can't make the constructor `constexpr`.

Without default initializer for `Buf`, the following error will be emitted: `defaulted definition of default constructor is not constexpr`. The reason for that is because `constexpr` constructors must have all fields initialized: [cppreference](https://en.cppreference.com/w/cpp/language/constexpr):

> **constexpr constructor**
> A `constexpr` constructor whose function body is not `=delete;` must satisfy the following additional requirements:
> - for the constructor of a class or struct, every base class sub-object and every non-variant non-static data member must be initialized. 

Apparently, trivial default constexpr constructors are only available in C++23: [P1331R2](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2019/p1331r2.pdf).

Not making the default constructor trivial brings several issues (of different impact):
- `sycl::bit_cast<sycl::half>` doesn't work anymore, because it requires `To` type to be trivially default constructible. I don't see any reason for such strict requirement and therefore I'm relaxing it in this PR
- putting `sycl::half` into `union` requires a bit more work from the user: from [cppreference](https://en.cppreference.com/w/cpp/language/union):
  > If a union contains a non-static data member with a non-trivial [default constructor](https://en.cppreference.com/w/cpp/language/default_constructor), the default constructor of the union is deleted by default unless a [variant member](https://en.cppreference.com/w/cpp/language/union#Union-like_classes) of the union has a default member initializer.
  
  That effectively means that user has to either provide default constructor for `union` containing `sycl::half` or provide default member initializer for `sycl::half` member to make it compile. You can find both approached used in this PR to adjust both implementation and tests.
- `device_global` currently requires trivially default constructible types, i.e. it means we can't have `device_global<sycl::half>`: https://github.com/intel/llvm/blob/3d8a136c0b1b230de998c7b34d716ececa868187/sycl/include/sycl/ext/oneapi/device_global/device_global.hpp#L128